packages/ripple/index.tsx

```diff
@@ -0,0 +1,319 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+import * as React from 'react';
+import * as classnames from 'classnames';
+import {Subtract} from 'utility-types'; // eslint-disable-line no-unused-vars
+// @ts-ignore
+import {MDCRippleFoundation, util} from '@material/ripple/dist/mdc.ripple';
+
+const MATCHES = util.getMatchesProperty(HTMLElement.prototype);
+
+interface RippledComponentProps<T> {
+  unbounded: boolean,
+  disabled: boolean,
+  style: React.CSSProperties,
+  className: string,
+  onMouseDown: React.MouseEventHandler<T>,
+  onMouseUp: React.MouseEventHandler<T>,
+  onTouchStart: React.TouchEventHandler<T>,
+  onTouchEnd: React.TouchEventHandler<T>,
+  onKeyDown: React.KeyboardEventHandler<T>,
+  onKeyUp: React.KeyboardEventHandler<T>,
+  onFocus: React.FocusEventHandler<T>,
+  onBlur: React.FocusEventHandler<T>,
+  computeBoundingRect?: (surface: T) => {},
+}
+
+interface RippledComponentState {
+  classList: Set<string>,
+  style: React.CSSProperties,
+}
+
+export interface InjectedProps<S, A = Element> extends RippledComponentProps<S> {
+  initRipple: (surface: S, activator?: A) => void,
+};
+
+function isElement(element: any): element is Element {
+  return element[MATCHES as 'matches'] !== undefined;
+}
+
+type ActivateEventTypes<Surface>
+  = React.MouseEvent<Surface> | React.TouchEvent<Surface> | React.KeyboardEvent<Surface> | React.FocusEvent<Surface>;
+
+const withRipple = <
+  P extends InjectedProps<Surface, Activator>,
+  Surface extends Element = Element,
+  Activator extends Element = Element
+>(
+    WrappedComponent: React.ComponentType<P>
+  ) => {
+  return class RippledComponent extends React.Component<
+  // Subtract removes any props "InjectedProps" if they are on "P"
+  // This allows the developer to override any props
+  // https://medium.com/@jrwebdev/react-higher-order-component-patterns-in-typescript-42278f7590fb
+    Subtract<P, InjectedProps<Surface, Activator>> & RippledComponentProps<Surface>,
+    RippledComponentState
+    > {
+    foundation_: MDCRippleFoundation | null = null;
```

Comment:
Are these private? If so, should we drop the trailing underscore and prefix them with `private`?

Reply:
Also, could this be an optional type instead of nullable?

Reply:
RE nullable v optional: ya!
RE private: But if I make it private I refer to it in the tests.

Reply:
You could declare it as private then add a public getter for both. Would that work or are you setting them in the tests, too?

Reply:
in some instances I do set the foundation for stubbing/spying. I know these should be private, so I was thinking we'd do another round of system changes to converting things to be private. Although adding the getter/setter doesn't sound too bad. What are your thoughts?

Reply:
I see you also have `initializeFoundation_` and `createAdapter_` which are public but should probably be private, too. Some ideas:
1. Make the `foundation` property configurable via the constructor
2. Make the constructor take an optional factory function that returns an implementation of `MDCRippleFoundation`. You can then update the constructor to set the private foundation property to the return value of the foundation factory OR initialize it with the default behavior.

I can mock up what 2 looks like if you want.

Reply:
Ugh I forgot this is React which means modifying the constructor is silly... You could add a `MDCRippleFoundationFactory` field to the `props` that accepts an optional function with the signature `(surface: Surface, activator?: Activator): MDCRippleFoundation`. Then you could pass configure that via props while still having the `foundation` and `unbounded` fields private.